proposals/eligibility_validator.go

```diff
@@ -74,35 +68,44 @@ func (v *Validator) CheckEligibility(ctx context.Context, ballot *types.Ballot,
 	if len(ballot.EligibilityProofs) == 0 {
 		return false, fmt.Errorf("empty eligibility list is invalid (ballot %s)", ballot.ID())
 	}
-	owned, err := v.cdb.GetAtxHeader(ballot.AtxID)
-	if err != nil {
-		return false, fmt.Errorf("failed to load atx %v: %w", ballot.AtxID, err)
-	}
-	if owned.TargetEpoch() != ballot.Layer.GetEpoch() {
-		return false, fmt.Errorf("%w: ATX target epoch (%v), ballot publication epoch (%v)",
-			errTargetEpochMismatch, owned.TargetEpoch(), ballot.Layer.GetEpoch())
-	}
-	if ballot.SmesherID != owned.NodeID {
-		return false, fmt.Errorf("%w: public key (%v), ATX node key (%v)", errPublicKeyMismatch, ballot.SmesherID.String(), owned.NodeID)
-	}
-	nonce, err := v.nonceFetcher.VRFNonce(ballot.SmesherID, ballot.Layer.GetEpoch())
-	if err != nil {
-		return false, fmt.Errorf("no vrf nonce for %v in epoch %v: %w", ballot.SmesherID, ballot.Layer.GetEpoch(), err)
-	}
-	var data *types.EpochData
-	if ballot.EpochData != nil && ballot.Layer.GetEpoch() == v.clock.CurrentLayer().GetEpoch() {
-		var err error
-		data, err = v.validateReference(ballot, actives, owned)
+	var atx *cache.ATXData
+	if epoch := ballot.Layer.GetEpoch(); v.cache.IsEvicted(epoch) {
```

Comment:
i find it weird that i have to care about whether this epoch is evicted from cache or not.
this check is also not atomic. if it's not evicted now, it might be evicted in the next microsecond.
also why i think it's better to have a db inside the cache, encapsulate the logic, and serve whatever the client is asking and decide for itself whether to cache in memory or not.

Reply:
this is very unlikely event, in such unlikely event code will have to retry. locking cache (to make it atomic) so that it can read from db is mistake